livre/eq/settings/CameraSettings.cpp

```diff
@@ -25,76 +25,84 @@
 
-    modelRotation_.pre_rotate_x( x );
-    modelRotation_.pre_rotate_y( y );
-    modelRotation_.pre_rotate_z( z );
-    setDirty( DIRTY_ALL );
+    modelview[0][3] = 0.0f;
+    modelview[1][3] = 0.0f;
+    modelview[2][3] = 0.0f;
+
+    modelview.pre_rotate_x( x );
+    modelview.pre_rotate_y( y );
+
+    modelview[0][3] = matrixValues[12];
+    modelview[1][3] = matrixValues[13];
+    modelview[2][3] = matrixValues[14];
+
+    std::copy( &modelview.array[0], &modelview.array[0] + 16, getMatrix( ));
 }
 
 void CameraSettings::moveCamera( const float x, const float y, const float z )
 {
-    cameraPosition_ += Vector3f( x, y, z );
-    setDirty( DIRTY_ALL );
+    getMatrix()[12] += x;
+    getMatrix()[13] += y;
+    getMatrix()[14] += z;
 }
 
-void CameraSettings::setCameraPosition( const Vector3f& position )
+void CameraSettings::setCameraPosition( const Vector3f& pos )
 {
-    cameraPosition_ = position;
-    setDirty( DIRTY_ALL );
+    getMatrix()[12] = pos.x();
+    getMatrix()[13] = pos.y();
+    getMatrix()[14] = pos.z();
 }
 
 void CameraSettings::setCameraLookAt( const Vector3f& lookAt )
 {
-    setModelViewMatrix( maths::computeModelViewMatrix( cameraPosition_, lookAt ));
+    const Vector3f eye( (float)getMatrix()[12],
+                        (float)getMatrix()[13],
+                        (float)getMatrix()[14]);
+    const Vector3f zAxis = vmml::normalize( eye - lookAt );
+
+    // Avoid Gimbal lock effect when looking upwards/downwards
+    Vector3f up( Vector3f::UP );
+    const float angle = zAxis.dot( up );
+    if( 1.f - std::abs( angle ) < 0.0001f )
+    {
+        Vector3f right( Vector3f::RIGHT );
+        if( angle > 0 ) // Looking downwards
+            right = Vector3f::LEFT;
+        up = up.rotate( 0.01f, right );
+        up.normalize();
+    }
+
+    Matrix4f modelview( eye, lookAt, up );
+    std::copy( &modelview.array[0], &modelview.array[0] + 16, getMatrix( ));
 }
 
-void CameraSettings::setModelViewMatrix( const Matrix4f& modelViewMatrix )
+Matrix4f CameraSettings::computeModelViewMatrix() const
 {
-    Matrix3f rotationMatrix;
-    maths::getRotationAndEyePositionFromModelView( modelViewMatrix,
-                                                   rotationMatrix,
-                                                   cameraPosition_ );
-    Matrix3f inverseRotation = Matrix4f::IDENTITY;
-    rotationMatrix.transpose_to( inverseRotation );
-    modelRotation_ = inverseRotation;
-    modelRotation_( 3, 3 ) = 1;
-    cameraPosition_ = inverseRotation * -cameraPosition_;
-
-    setDirty( DIRTY_ALL );
-}
+    float matrixValues[16];
+    std::copy( getMatrix(), getMatrix() + 16, matrixValues );
 
-Matrix4f CameraSettings::getModelViewMatrix() const
-{
-    Matrix4f modelView;
-    modelView = modelRotation_;
-    modelView.set_translation( cameraPosition_ );
-    return modelView;
+    Matrix4f modelview( &matrixValues[0], &matrixValues[0] + 16 );
+    return modelview;
```

Comment:
Yes, it's using the Matrix( begin, end ) iterator ctor. pointers are iterators :)